README.md
# AT91Bootstrap 3.99.x (Glasnost Edition)

This repository is a fork of the original [AT91Bootstrap](https://github.com/linux4sam/at91bootstrap) project by Microchip, adapted for the Glasnost boards and modernized with a new build system and versioning scheme.

> **Note:** This fork was created because the 3.x series is no longer supported by upstream.

## What's Changed

- **Modern Build System:** The build system has been reworked for improved maintainability and ease of use.
- **Glasnost Board Support:** Fixes and enhancements for reliable operation on Glasnost hardware.
- **New Versioning:** Releases now follow the `3.99.x` scheme to distinguish from upstream and indicate ongoing development.

## About AT91Bootstrap

AT91Bootstrap is a 2nd level bootloader for Microchip (formerly Atmel) AT91 microprocessors. It handles hardware initialization (clocks, PIO, DRAM, etc.) and loads the main application from various boot media (NAND, DataFlash, EEPROM, SD Card, etc.) into main memory.

For full documentation, usage instructions, and supported boards, see the [official AT91Bootstrap repository](https://github.com/linux4sam/at91bootstrap) and [project wiki](https://www.linux4sam.org/bin/view/Linux4SAM/AT91Bootstrap).

## Attribution

This project is based on the original work by Microchip and the AT91Bootstrap community.  
See [LICENSE](LICENSE) for details.

Original authors and maintainers [AT91Bootstrap contributors](https://github.com/linux4sam/at91bootstrap/graphs/contributors)





ChangeLog
AT91Bootstrap ChangeLog 
	
Version 3.4.5


	Jan 11 09:43:39 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* Only for at91sam9260ek nand booting, using "-Os" instead of using "-O2"
	  as CPPFLAGS option. Since using 2010q1 toolchain, the binary size built
	  with "at91sam9260eknf_uboot_defconfig" is greater the maxium size (4096).

	Jan 8 10:45:56 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* When failed to load the 1-Wire information, print WARNING message,
	  then continue to go instead of hang.

	Jan 7 16:56:06 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* When loading uboot, the "CONFIG_JUMP_ADDR" 0x23F00000 ==> 0x21F00000,
	  and the offset of nandflash, "CONFIG_IMG_ADDRESS" 0x20000 ==>0x40000.

	Jan 9 14:46:09 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* When diabling ""CONFIG_DEBUG", i.e. "dbg_log" doesn't work,
	  using "dbgu_print" print some simple promt string message.

	Jan 7 14:51:59 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* It always configis dbgu pins during board initialization
	  despite of "CONFIG_DEBUG".

	Jan 7 13:12:20 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* [BUG] Using 2009q3, 2010q1 and buildroot toolchain
	  to build at91sam9g10ek with "at91sam9g10eknf_uboot_defconfig".
	  It print out "Bad Block#xxx" for all nand blocks.

	Jan 7 10:38:03 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* When ChkFileSize greater than BOOTSTRAP_MAXSIZE,
	  it will not produce the binary.

	Jan 6 16:01:23 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* Set the machine type number to 0xffffffff
	  when loading the kernel image and dt.

	Jan 4 11:40:13 2013 Wenyou Yang <wenyou.yang@atmel.com>

	* [BUG] Fix after erasing block0 the status register
	  reports erasing failure, even if the erasing is done.

	Dec 17 16:25:40 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Add the 1-Wire support for sama5d3xek

	Nov 6 14:50:14 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Add DT support for loading kernel image.

	Nov 26 13:05:41 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Since the GPL violation, rewrite the mci & sd driver code
	  without GPL code.

	Nov 20 11:15:27 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Since the GPL violation, rewrite the spi & spi_flash driver
	  code without GPL code.

	Nov 26 15:04:36 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Rewrite the kernel image loading code,
	  Since the GPL violation, rewrite the kernel image loading code

	Dec 5 16:58:18 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Add support to config disable the watchdog

	Dec 7 19:13:01 2012 Josh Wu <josh.wu@atmel.com>

	* Add support to 4-bits ECC

	Dec 4 17:36:58 2012 Josh Wu <josh.wu@atmel.com>

	* Port the PMECC correct data function from SAM-BA code
	  Before, using this function exported from the ROMCode.
	  Since the new revision of SAMA5, this function cannot
	  be accessed from the ROMCode. So porting the PMECC
	  correct data function from SAM-BA.

	Oct 31 10:45:39 2012 Wenyou Yang <wenyou.yang@atmel.com>

	* Fix [BUG] Nand recovery function not erase the block0


Version 3.4.4

	2012-09-24  Wenyou Yang  <wenyou.yang@atmel.com>

	* Align to the U-Boot latest version, change the default 
	  jump address from 0x23f00000 to 0x21f00000 
	  for U-Boot in the external memory to the boards whose 
	  memory size is 64M. These boards include, at91sam9260ek,
	  at91sam9261ek, at91sam9263ek, at91sam9g10ek, 
	  at91sam9g20ek, and at91sam9rlek.

	2012-09-18  Wenyou Yang  <wenyou.yang@atmel.com>

	* Workaround for HDMI and QT1070 share irq line.

	* Disable the nandflash chip internal ecc function 
	  for some Micron chip.

	* Fix the bug that EMMC does not work

	2012-09-17  Bo Shen  <voice.shen@atmel.com>

	* Add one wire information for PDA DM module for 
	  at91sam9x5ek board.

	2012-09-05  Wenyou Yang  <wenyou.yang@atmel.com>
	
	* Enable the SLOW slew rate for SDRAM clock for 
	  at91sam9g20ek board.


Version 3.4.3
	
	2012-08-22  Wenyou Yang  <wenyou.yang@atmel.com>

	* Add the revision

	2012-08-13  Wenyou Yang  <wenyou.yang@atmel.com>

	* Change the default CONFIG_IMG_SIZE for U-Boot
	  from 0x50000 to 0x80000
	 
	* Update the default config file for at91sam9x5ek 
	  and at91sama5d3xek boards

	* Update the PMECC lookup table offset information
	  in the ROM.

	2012-08-01  Wenyou Yang  <wenyou.yang@atmel.com>
	
	* Add support to automatically generation of ecc layout 
	  for software ecc

	2012-07-25  Wenyou Yang  <wenyou.yang@atmel.com>

	* Update the sam9260 dedicated nandflash driver code

	* Workaround for pmecc to deal with All-0xFF page.

	* Add support to 4K page nandflash

	2012-07-24  Wenyou Yang  <wenyou.yang@atmel.com>

	* Only do checking the bad block at starting to read a new block,
	  instead of at reading every page.
	
	* Fix the bug that PMECC module does not work while reading 
	  data from nandflash.

	* Change the mode of sending address for nandflash command
	  to get more flexible.

	* Add the code to automatically generate the ecc layout 
	  according to the nandflash information, instead of 
	  the static definiton for ooblayout for ecc data


Version 3.4.2

	2012-07-11  Nicolas Ferre  <nicolas.ferre@atmel.com>

	* Remove the sddrc driver code as it is not 
	  needed anymore.

	2012-07-11  Wenyou Yang  <wenyou.yang@atmel.com>

	* Change the implement of udelay() function 
	  with the PIT timer to get more accurate time delay, 
	  independent on CPU frequency.

	* Remove the delay() function which implement depend on 
	  CPU frequency.

	2012-07-11  Nicolas Ferre  <nicolas.ferre@atmel.com>

	* Modify timings for 512MB DDR2 on at91sama5d3xek board

	* Better presentation of value for io calibration and 
	  DDR DLL Master/Slave offsets
	 
	* Modify t1pr to ease understanding
	
	2012-07-10  Nicolas Ferre  <nicolas.ferre@atmel.com>

	* Remove the sw recovery feature from configuration of
	  at91sama5d3xek board

	* Remove pull-ups in MCI pin configuration on at91sama5d3xek
	  board.


Version 3.4.1

	2012-07-05  Wenyou Yang  <wenyou.yang@atmel.com>


fs/diskio.c
/*-----------------------------------------------------------------------*/
/* Low level disk I/O module skeleton for FatFs     (C)ChaN, 2007        */
/*-----------------------------------------------------------------------*/
/* This is a stub disk I/O module that acts as front end of the existing */
/* disk I/O modules and attach it to FatFs module with common interface. */
/*-----------------------------------------------------------------------*/
#include "ff/diskio.h"
#include "ff/ffconf.h"
#include "ff/integer.h"
#include "media.h"

//------------------------------------------------------------------------------
//         Internal variables

static volatile DSTATUS Stat = STA_NOINIT;	/* Disk status */

//------------------------------------------------------------------------------
/* Initialize a Drive                                                    */
/*-----------------------------------------------------------------------*/

DSTATUS disk_initialize(BYTE drv        /* Physical drive number (0..) */
    )
{
	if (drv) return STA_NOINIT;	
	
	if (mmc_initialize() == 0)
		Stat &= ~STA_NOINIT;

	return Stat;
}

/*-----------------------------------------------------------------------*/
/* Return Disk Status                                                    */
/*-----------------------------------------------------------------------*/

DSTATUS disk_status(BYTE drv    /* Physical drive number (0..) */
    )
{
	if (drv) return STA_NOINIT;

	return Stat;
}

/*-----------------------------------------------------------------------*/
/* Read Sector(s)                                                        */
/*-----------------------------------------------------------------------*/

DRESULT disk_read(BYTE drv,     /* Physical drive number (0..) */
                  BYTE *buff,  /* Data buffer to store read data */
                  DWORD sector, /* Start sector number (LBA) */
                  BYTE count    /* Sector count (1..255) */
    )
{
	if (drv || !count) return RES_PARERR;
	if (Stat & STA_NOINIT) return RES_NOTRDY;

	if (mmc_bread((unsigned int)sector, (unsigned int)count, (void *)buff) == count)
		return RES_OK;
	else
		return RES_ERROR;
}

/*-----------------------------------------------------------------------*/
/* Write Sector(s)                                                       */
/*-----------------------------------------------------------------------*/

#if _READONLY == 0
DRESULT disk_write(BYTE drv,    /* Physical drive number (0..) */
                   const BYTE * buff,   /* Data to be written */
                   DWORD sector,        /* Sector number (LBA) */
                   BYTE count   /* Sector count (1..255) */
    )
{
}
#endif  /* _READONLY */

/*-----------------------------------------------------------------------*/
/* Miscellaneous drive controls other than data read/write               */
/*-----------------------------------------------------------------------*/
#if _USE_IOCTL == 1
DRESULT disk_ioctl(BYTE drv,    /* Physical drive number (0..) */
                   BYTE ctrl,   /* Control code */
                   void *buff   /* Buffer to send/receive control data */
    )
{
}
#endif

/*----------------------------------------------------------------------*/
/* User Provided RTC Function for Fats module				*/
/*----------------------------------------------------------------------*/
/* Currnet time is returned with packed into a DWORD value.		*/
/* The bit field is as follows:						*/
/*   bit31:25  Year from 1980 (0..127)					*/
/*   bit24:21  Month (1..12)						*/
/*   bit20:16  Day in month(1..31)					*/
/*   bit15:11  Hour (0..23)						*/
/*   bit10:5   Minute (0..59)						*/
/*   bit4:0    Second / 2 (0..29)					*/

DWORD get_fattime(void)
{
    DWORD time;

    time = ((2009 - 1980) << 25)
        | (9 << 21)
        | (15 << 16)
        | (17 << 11)
        | (45 << 5)
        | ((59 / 2) << 0);

    return time;
}
